google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1beta2/StreamWriterV2.java

```diff
@@ -0,0 +1,352 @@
+      }
+      if (connectionFinalStatus != null) {
+        requestWrapper.appendResult.setException(
+            new StatusRuntimeException(
+                Status.fromCode(Status.Code.FAILED_PRECONDITION)
+                    .withDescription(
+                        "Stream is closed due to " + connectionFinalStatus.toString())));
+        return requestWrapper.appendResult;
+      }
+      waitingRequestQueue.addLast(requestWrapper);
+      hasMessageInWaitingQueue.signal();
+      return requestWrapper.appendResult;
+    } finally {
+      this.lock.unlock();
+    }
+  }
+
+  /** Close the stream writer. Shut down all resources. */
+  @Override
+  public void close() {
+    log.info("User closing stream: " + streamName);
+    this.lock.lock();
+    try {
+      this.userClosed = true;
+    } finally {
+      this.lock.unlock();
+    }
+    log.info("Waiting for append thread to finish. Stream: " + streamName);
+    try {
+      appendThread.join();
+      log.info("User close complete. Stream: " + streamName);
+    } catch (InterruptedException e) {
+      // Unexpected. Just swallow the exception with logging.
+      log.warning(
+          "Append handler join is interrupted. Stream: " + streamName + " Error: " + e.toString());
+    }
+  }
+
+  /*
+   * This loop is executed in a separate thread.
+   *
+   * It takes requests from waiting queue and sends them to server.
+   */
+  private void appendLoop() {
+    Deque<AppendRequestAndResponse> localQueue = new LinkedList<AppendRequestAndResponse>();
+    while (!waitingQueueDrained()) {
+      this.lock.lock();
+      try {
+        hasMessageInWaitingQueue.await(100, TimeUnit.MILLISECONDS);
+        while (!this.waitingRequestQueue.isEmpty()) {
+          localQueue.addLast(this.waitingRequestQueue.pollFirst());
+        }
+      } catch (InterruptedException e) {
+        log.warning("Interrupted while waiting for message. Error: " + e.toString());
+      } finally {
+        this.lock.unlock();
+      }
+
+      if (localQueue.isEmpty()) {
+        continue;
+      }
+
+      // TODO: Add reconnection here.
+
+      this.lock.lock();
+      try {
+        while (!localQueue.isEmpty()) {
+          AppendRequestAndResponse requestWrapper = localQueue.pollFirst();
+          this.inflightRequestQueue.addLast(requestWrapper);
+          this.streamConnection.send(requestWrapper.message);
+        }
+      } finally {
+        this.lock.unlock();
+      }
+    }
+
+    log.info("Cleanup starts. Stream: " + streamName);
+    // At this point, the waiting queue is drained, so no more requests.
+    // We can close the stream connection and handle the remaining inflight requests.
+    this.streamConnection.close();
+
+    log.info("Waiting for done callback from stream connection. Stream: " + streamName);
+    long waitDeadlineMs = System.currentTimeMillis() + DONE_CALLBACK_WAIT_TIMEOUT.toMillis();
+    while (true) {
+      if (System.currentTimeMillis() > waitDeadlineMs) {
```

Comment:
we can't have a real deadline here, we must wait for the done callback, otherwise the done callback could be called at any time (and may cause exception).

To make the code more readable, consider a small member function such as WaitForStreamConnectionFinished or similar around the done callback, so here we just call that function.

Reply:
Done.